### PR TITLE
Add a plugin mechanism to tukit

### PR DESCRIPTION
**Pull request by aplanas**

### CODE REVIEW

File: lib/Transaction.hpp [hunk 1]
```diff
@@ -167,6 +167,14 @@ class Transaction {
...
      */
     std::filesystem::path getRoot();
+
+    /**
+     * @brief Return the bindDir path of the snapshot
+     * @return bindDir path
+     *
+     * The bindDir is the live snapshot, that includes auxiliary mounts as /etc or /run.
+     */
+    std::filesystem::path getBindDir();
```
Comment (aplanas) [inline lib/Transaction.hpp:177]: Not happy about this, but the alternatives that I thought were worse. Any suggestion?

Comment (aplanas) [reply lib/Transaction.hpp:172]: Resolved in the next commit

**Comment by aplanas** [comment]
Ready for reviews!

cc @Vogtinator

**Comment by aplanas** [comment]
@laenion @Vogtinator ping? : )

File: lib/Plugins.cpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    std::string output;
+
+    for (auto& p: plugins) {
+        std::string cmd = p.string() + " " + stage;
+        if (!args.empty())
+            cmd.append(" " + args);
+
+        try {
+            output = Util::exec(cmd);
+            tulog.info("plugin ", p, " output: ", output);
```
Comment (laenion) [inline lib/Plugins.cpp:67]: I'm not sure we want to print this text every time, even when when the plugin didn't even print something itself. On the other hand I do indeed see the necessity that some plugins have to print important information. I'd suggest to only print this line if the output is not empty.

A second point: When starting tukit with `--verbose`, then the output of the plugin will be printed by `Util::exec` itself already, but I guess it's acceptable when the output it printed twice then.

And last: Please use a capital 'P' in "plugin", or - which I'd prefer - "Output of plugin xxx: "

File: lib/Plugins.cpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        } catch (const ExecutionException &e) {
+            // An error in the plugin should not discard the transaction
+            tulog.error("ERROR: plugin ", p, " fails with ", e.what());
```
Comment (laenion) [inline lib/Plugins.cpp:70]: Capital "P" in "plugin
"fails" -> "failed"

File: tukit/tukit.cpp [hunk 4]
```diff
@@ -11,6 +11,7 @@
 #include "Transaction.hpp"
 #include "Reboot.hpp"
 #include "Log.hpp"
+#include "Plugins.hpp"
```
Comment (laenion) [inline tukit/tukit.cpp:14]: I assume the plugins are also supposed to run when calling tukit via D-Bus or the C and C++ APIs. Due to that all of the calls from below have to be moved to the functions they are wrapping.

File: sbin/Makefile.am [hunk 5]
```diff
@@ -12,6 +12,7 @@ EXTRA_DIST = transactional-update.in
 
 do_subst = sed -e 's,[@]sysconfdir[@],$(sysconfdir),g' \
 	-e 's,[@]prefix[@],$(prefix),g' \
+	-e 's,[@]libdir[@],$(libdir),g' \
```
Comment (laenion) [inline sbin/Makefile.am:15]: This seems to be unused?

**Review by laenion — CHANGES_REQUESTED** [review]
In general I like the approach - see the nits, but the place to call the plugins will have to be changed in any case.

(I'm still not sure whether it's a good idea to have the BindDir as part of the API, but I'll not complain for now ;-))

File: doc/tukit-plugins.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+to `/dev/null`.  This will be used as a mark to completely disable
+this plugin, and would not be called by `tukit`.
+
+The plugins in `/etc` will be called before the ones in `/usr` but the
+user should not depend on the calling order.
+
+## Stages
+
+Some actions will trigger a plugin call before and after the action
+itself, depending if makes sense in the context of this action.  The
```
[the remark this answers is not in the document]
Comment (aplanas) [reply doc/tukit-plugins.md:51]: done

File: doc/tukit-plugins.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+next table summarize the action, the stage and different parameters
```
[the remark this answers is not in the document]
Comment (aplanas) [reply doc/tukit-plugins.md:52]: done

File: doc/tukit-plugins.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+```bash
+#!/bin/bash
+
+call_pre() {
+	local path="$1"; shift
+	local snapshot_id="$1"; shift
+	local cmd="$@"
+
+    # The live snapshot is in "$path", and the future closed snapshot in
+	# "/.snapshots/${snapshot_id}/snapshot
```
[the remark this answers is not in the document]
Comment (aplanas) [reply doc/tukit-plugins.md:93]: done

File: lib/Plugins.cpp [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+            if (filesystem::is_symlink(path) && filesystem::read_symlink(path) == "/dev/null") {
+                plugins_set.insert(filename);
+                continue;
+            }
+
+            // If the plugin is not executable, ignore it
+            if (!(filesystem::is_regular_file(path) && (access(path.c_str(), X_OK) == 0)))
+                continue;
+
+            tulog.info("Regiter plugin ", path);
```
[the remark this answers is not in the document]
Comment (aplanas) [reply lib/Plugins.cpp:46]: yes, done

File: lib/Plugins.cpp (hunk 2 above)
Comment (aplanas) [reply lib/Plugins.cpp:67]: done

File: lib/Plugins.cpp (hunk 3 above)
Comment (aplanas) [reply lib/Plugins.cpp:70]: done

File: sbin/Makefile.am (hunk 5 above)
Comment (aplanas) [reply sbin/Makefile.am:15]: I think that is used in transactional-update.in, in variable STATUS_FILE, BASE_MANIFEST_FILE and SYSTEM_MANIFEST_FILE?

File: tukit/tukit.cpp (hunk 4 above)
Comment (aplanas) [reply tukit/tukit.cpp:14]: That is a good point, but they cannot be moved into the commands?

There is not a perfect mapping, for example, the abort() cannot be mapped intro resume(), as it is used also in close()

Comment (aplanas) [reply tukit/tukit.cpp:14]: I guess that the best option is to also add the plugins into the bindings and dbus entry points?, or do you see a way to map then?

Comment (aplanas) [reply tukit/tukit.cpp:14]: ah, cannot be added into the binding, a binding should be created instead ...

**Comment by aplanas** [comment]
@laenion should be ready for another pass

File: sbin/Makefile.am (hunk 5 above)
Comment (laenion) [reply sbin/Makefile.am:15]: Indeed. That may have been one reason why the `status` command hasn't worked as intended...

File: doc/tukit-plugins.md [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+| Action   | Stage | Parameters                        | Notes |
+|----------|-------|-----------------------------------|-------|
+| init     | -pre  |                                   |       |
+|          | -post | path, snapshot\_id                |       |
+| resume   | -pre  | snapshot\_id                      |       |
+|          | -post | path, snapshot\_id                |       |
+| execute  | -pre  | path, snapshot\_id, action params |       |
+|          | -post | path, snapshot\_id, action params |       |
+| callExt  | -pre  | path, snapshot\_id, action params | [1]   |
+|          | -post | path, snapshot\_id, action params |       |
```
Comment (laenion) [inline doc/tukit-plugins.md:70]: I'm wondering whether it wouldn't make sense to also have one "common" action target such as `runCommand` to run a plugin independent of the specific entry point...
Unfortunately naming is not ideal :/
Opinions?

File: doc/tukit-plugins.md [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+to `/dev/null`.  This will be used as a mark to completely disable
+this plugin, and would not be called by `tukit`.
+
+The plugins in `/etc` will be called before the ones in `/usr` but the
+user should not depend on the calling order.
+
+## Stages
+
+The actions are linked to the low-level API from libtukit, not from
+the user API level.  This means that some verbsq like `tukit exeture
```
[the remark this answers is not in the document]
Comment (aplanas) [reply doc/tukit-plugins.md:51]: yes, done

File: doc/tukit-plugins.md (hunk 10 above)
Comment (aplanas) [reply doc/tukit-plugins.md:70]: For me is more important the consistency. If we are mapping the libtukit, we need to reflect it in the plugins.

Comment (aplanas) [reply doc/tukit-plugins.md:70]: uhmm also there will be two entry points for the same action, that will be very confusing ...

File: doc/tukit-plugins.md [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+| execute  | -pre  | path, snapshot\_id, action params |       |
+|          | -post | path, snapshot\_id, action params |       |
+| callExt  | -pre  | path, snapshot\_id, action params | [1]   |
+|          | -post | path, snapshot\_id, action params |       |
+| finalize | -pre  | path, snapshot\_id                |       |
+|          | -post | snapshot\_id, [discarded]         | [2]   |
+| abort    | -post | snapshot\_id                      | [3]   |
+| keep     | -pre  | path, snapshot\_id                |       |
+|          | -post | snapshot\_id                      |       |
+
```
[the remark this answers is not in the document]
Comment (aplanas) [reply doc/tukit-plugins.md:76]: Added the `reboot` one.

**Comment by laenion** [comment]
Thanks a lot for all your work, everything is looking good now! Just one last thing: The last two commits are messed up, "[tukit: move getBindDir as protected](https://github.com/openSUSE/transactional-update/pull/122/commits/857c8969fec2811545d07262b55ff4a54b9995a0)" now contains all the changes that should have gone into "[tukit: add plugin mechanism](https://github.com/openSUSE/transactional-update/pull/122/commits/c374ee94d6684efa5adeda71d55e94fe698671a3)"

**Comment by aplanas** [comment]
> The last two commits are messed up

Ah indeed. I squashed both.

**Review by laenion — COMMENTED** [review]
Looking good now, thanks a lot for your work!

**Comment by aplanas** [comment]
Thanks for your reviews @laenion. That helped a lot.

**Pull request merged**